### PR TITLE
ARROW-6869: [C++] Do not return invalid arrays from DictionaryBuilder::Finish when reusing builder. Add "FinishDelta" method and "ResetFull" method

### PR DESCRIPTION
**Pull request by wesm**
The current behavior is "wrong enough" that adding this to 0.15.1 might not be a bad idea, but we could always leave it for the next major release

### CODE REVIEW

**Comment by wesm** [comment]
cc @alendit also since this touches code that you worked on

File: cpp/src/arrow/array_dict_test.cc [hunk 1]
```diff
@@ -336,9 +320,34 @@ TYPED_TEST(TestDictionaryBuilder, FinishResetBehavior) {
...
+  ASSERT_OK(builder.Append(static_cast<c_type>(2)));
+
+  std::shared_ptr<Array> result;
+  ASSERT_OK(builder.Finish(&result));
+
+  ASSERT_OK(builder.Append(static_cast<c_type>(3)));
+  ASSERT_OK(builder.Finish(&result));
+
+  // Dictionary expanded
+  ASSERT_EQ(3, static_cast<const DictionaryArray&>(*result).dictionary()->length());
```
Comment (pitrou) [inline cpp/src/arrow/array_dict_test.cc]: Can we actually test the result value here?

File: cpp/src/arrow/array_dict_test.cc [hunk 2]
```diff
@@ -336,9 +320,34 @@ TYPED_TEST(TestDictionaryBuilder, FinishResetBehavior) {
...
+
+  builder.ResetFull();
+  ASSERT_OK(builder.Append(static_cast<c_type>(4)));
+  ASSERT_OK(builder.Finish(&result));
+  ASSERT_EQ(1, static_cast<const DictionaryArray&>(*result).dictionary()->length());
```
Comment (pitrou) [inline cpp/src/arrow/array_dict_test.cc]: Same here.

**Review by pitrou — COMMENTED** [review]
The API changes look good. Perhaps just enhance the tests a bit.

File: cpp/src/arrow/array/builder_dict.h [hunk 3]
```diff
@@ -230,10 +226,16 @@ class DictionaryBuilderBase : public ArrayBuilder {
   }
 
   void Reset() override {
+    // Perform a partial reset. Also call ResetDictionary to reset the
```
[the remark this answers is not in the document]
Comment (wesm) [reply cpp/src/arrow/array/builder_dict.h]: Yes, fixed

File: cpp/src/arrow/array/builder_dict.h [hunk 4]
```diff
@@ -244,28 +246,26 @@ class DictionaryBuilderBase : public ArrayBuilder {
...
 
-  Status FinishInternal(std::shared_ptr<ArrayData>* out) override {
-    // Finalize indices array
-    ARROW_RETURN_NOT_OK(indices_builder_.FinishInternal(out));
+  /// \brief Return dictionary indices and a delta dictionary since the last
+  /// time that Finish or FinishDelta were called, and reset state of builder
+  /// (except the memo table)
+  Status FinishDelta(std::shared_ptr<Array>* out_indices,
+                     std::shared_ptr<Array>* out_delta) {
+    std::shared_ptr<Array> dictionary;
```
[the remark this answers is not in the document]
Comment (wesm) [reply cpp/src/arrow/array/builder_dict.h]: fixed

File: cpp/src/arrow/array/builder_dict.h [hunk 5]
```diff
@@ -244,28 +246,26 @@ class DictionaryBuilderBase : public ArrayBuilder {
...
 
-    // Generate dictionary array from hash table contents
-    std::shared_ptr<ArrayData> dictionary_data;
+    *out_indices = MakeArray(indices_data);
+    return Status::OK();
+  }
 
-    ARROW_RETURN_NOT_OK(
-        memo_table_->GetArrayData(pool_, delta_offset_, &dictionary_data));
+  Status FinishInternal(std::shared_ptr<ArrayData>* out) override {
```
[the remark this answers is not in the document]
Comment (wesm) [reply cpp/src/arrow/array/builder_dict.h]: yes, fixed

File: cpp/src/arrow/array_dict_test.cc (hunk 1 above)
Comment (wesm) [reply cpp/src/arrow/array_dict_test.cc]: done

File: cpp/src/arrow/array_dict_test.cc (hunk 2 above)
Comment (wesm) [reply cpp/src/arrow/array_dict_test.cc]: done

**Review by wesm — COMMENTED** [review]
Thanks -- fixed the issues you cited

**Review by pitrou — APPROVED** [review]
+1 from me now.

**Comment by wesm** [comment]
I've marked this with 0.15.1 (since it's buggy) but if others disagree about including it, I am OK with that

**Pull request closed without merging**